signing/secp256k1/impl/src/main/java/tech/pegasys/signers/secp256k1/multikey/MultiKeySignerProvider.java

```diff
@@ -33,75 +36,95 @@
+    return new MultiKeySignerProvider(
+        signingMetadataTomlConfigLoader, azureFactory, hashicorpSignerFactory, configFileSelector);
   }
 
-  public MultiKeyTransactionSignerProvider(
+  public MultiKeySignerProvider(
       final SigningMetadataTomlConfigLoader signingMetadataTomlConfigLoader,
-      final AzureKeyVaultTransactionSignerFactory azureFactory,
-      final HashicorpSignerFactory hashicorpSignerFactory) {
+      final AzureKeyVaultSignerFactory azureFactory,
+      final HashicorpSignerFactory hashicorpSignerFactory,
+      final FileSelector<PublicKey> configFileSelector) {
     this.signingMetadataTomlConfigLoader = signingMetadataTomlConfigLoader;
     this.azureFactory = azureFactory;
     this.hashicorpSignerFactory = hashicorpSignerFactory;
+    this.configFileSelector = configFileSelector;
   }
 
   @Override
-  public Optional<TransactionSigner> getSigner(final String address) {
+  public Optional<Signer> getSigner(final PublicKey publicKey) {
     return signingMetadataTomlConfigLoader
-        .loadMetadataForAddress(address)
+        .loadMetadata(configFileSelector.getSpecificConfigFileFilter(publicKey))
         .map(metadataFile -> metadataFile.createSigner(this));
   }
 
   @Override
-  public Set<String> availableAddresses() {
-    return signingMetadataTomlConfigLoader.loadAvailableSigningMetadataTomlConfigs().stream()
-        .map(metadataFile -> metadataFile.createSigner(this))
+  public Set<PublicKey> availablePublicKeys() {
+    return signingMetadataTomlConfigLoader
+        .loadAvailableSigningMetadataTomlConfigs(configFileSelector.getAllConfigFilesFilter())
+        .stream()
+        .map(
+            metadataFile -> {
+              final Signer signer = metadataFile.createSigner(this);
+              try {
+                if ((signer != null)
+                    && configFileSelector
+                        .getSpecificConfigFileFilter(signer.getPublicKey())
+                        .accept(Path.of(metadataFile.getFilename()))) {
+                  return signer;
+                }
+                return null;
+              } catch (final IOException e) {
+                LOG.warn("IO Exception raised while loading {}", metadataFile.getFilename());
+                return null;
+              }
+            })
         .filter(Objects::nonNull)
-        .map(TransactionSigner::getAddress)
+        .map(Signer::getPublicKey)
         .collect(Collectors.toSet());
   }
 
   @Override
-  public TransactionSigner createSigner(final AzureSigningMetadataFile metadataFile) {
+  public Signer createSigner(final AzureSigningMetadataFile metadataFile) {
     try {
       return azureFactory.createSigner(metadataFile.getConfig());
     } catch (final TransactionSignerInitializationException e) {
-      LOG.error("Failed to construct Azure signer from " + metadataFile.getBaseFilename());
+      LOG.error("Failed to construct Azure signer from " + metadataFile.getFilename());
```

Comment:
add `e` as argument to error to get the whole stacktrace (or print actual error message from `TransactionSignerInitializationException` at least?).

Reply:
`TransactionSignerInitializationException` should be renamed to `SignerInitializationException` ??

Reply:
Renamed exception, I'm wondering if this is one of the places we deliberately didn't want the full stack trace (due to noise) - that maybe(?) this message provides sufficient information to resolve the issue.